From c602fa951f06c4ae8b3fa4653e545fecd3e87193 Mon Sep 17 00:00:00 2001
From: Milton <mingjuwu@google.com>
Date: Thu, 25 Sep 2025 07:23:07 +0000
Subject: [PATCH] Ignore enrollAfterFace in FaceEnroll from ext pkg

Remove key of enrollAfterFace in FaceEnroll if this intent is coming
from external package

Bug: 433252423
Flag: EXEMPT BUGFIX
Test: atest FaceEnrollTest
Change-Id: Ibbb6340163613ea955516f11579d767b2ade4312
(cherry picked from commit 7ac22f2eae5d3e8e65bc7f8db5a173231ebba071)
---
 .../settings/biometrics/face/FaceEnroll.kt    | 24 +++++++------
 .../biometrics/face/FaceEnrollTest.kt         | 36 ++++++++++++-------
 2 files changed, 38 insertions(+), 22 deletions(-)

diff --git a/src/com/android/settings/biometrics/face/FaceEnroll.kt b/src/com/android/settings/biometrics/face/FaceEnroll.kt
index 48ea527c03..40d3f08e76 100644
--- a/src/com/android/settings/biometrics/face/FaceEnroll.kt
+++ b/src/com/android/settings/biometrics/face/FaceEnroll.kt
@@ -23,10 +23,11 @@ import android.util.Log
 import androidx.annotation.VisibleForTesting
 import androidx.appcompat.app.AppCompatActivity
 import com.android.settings.biometrics.BiometricEnrollBase.RESULT_FINISHED
+import com.android.settings.biometrics.MultiBiometricEnrollHelper.EXTRA_ENROLL_AFTER_FACE
 import com.android.settings.biometrics.combination.CombinedBiometricStatusUtils
 import com.android.settings.overlay.FeatureFactory.Companion.featureFactory
 
-class FaceEnroll: AppCompatActivity() {
+class FaceEnroll : AppCompatActivity() {
 
     /**
      * The class of the next activity to launch. This is open to allow subclasses to provide their
@@ -39,8 +40,7 @@ class FaceEnroll: AppCompatActivity() {
     private val enrollActivityProvider: FaceEnrollActivityClassProvider
         get() = featureFactory.faceFeatureProvider.enrollActivityClassProvider
 
-    @VisibleForTesting
-    var launchedFromProvider: () -> String? = { launchedFromPackage }
+    @VisibleForTesting var launchedFromProvider: () -> String? = { launchedFromPackage }
 
     private var isLaunched = false
 
@@ -53,9 +53,9 @@ class FaceEnroll: AppCompatActivity() {
 
         if (!isLaunched) {
             /**
-             *  Logs the next activity to be launched, creates an intent for that activity,
-             *  adds flags to forward the result, includes any existing extras from the current intent,
-             *  starts the new activity and then finishes the current one
+             * Logs the next activity to be launched, creates an intent for that activity, adds
+             * flags to forward the result, includes any existing extras from the current intent,
+             * starts the new activity and then finishes the current one
              */
             Log.d("FaceEnroll", "forward to $nextActivityClass")
             val nextIntent = Intent(this, nextActivityClass)
@@ -63,6 +63,7 @@ class FaceEnroll: AppCompatActivity() {
 
             // drop extras that are not allowed from external packages before launching
             if (launchedFromProvider() != packageName) {
+                nextIntent.removeExtra(EXTRA_ENROLL_AFTER_FACE)
                 nextIntent.removeExtra(Intent.EXTRA_USER_ID)
             }
             startActivityForResult(nextIntent, 0)
@@ -80,13 +81,16 @@ class FaceEnroll: AppCompatActivity() {
         requestCode: Int,
         resultCode: Int,
         data: Intent?,
-        caller: ComponentCaller
+        caller: ComponentCaller,
     ) {
         super.onActivityResult(requestCode, resultCode, data, caller)
         isLaunched = false
-        if (intent.getBooleanExtra(
-                CombinedBiometricStatusUtils.EXTRA_LAUNCH_FROM_SAFETY_SOURCE_ISSUE, false)
-            && resultCode != RESULT_FINISHED) {
+        if (
+            intent.getBooleanExtra(
+                CombinedBiometricStatusUtils.EXTRA_LAUNCH_FROM_SAFETY_SOURCE_ISSUE,
+                false,
+            ) && resultCode != RESULT_FINISHED
+        ) {
             featureFactory.biometricsFeatureProvider.notifySafetyIssueActionLaunched()
         }
         setResult(resultCode, data)
diff --git a/tests/robotests/src/com/android/settings/biometrics/face/FaceEnrollTest.kt b/tests/robotests/src/com/android/settings/biometrics/face/FaceEnrollTest.kt
index ad4605aff9..b5ea747862 100644
--- a/tests/robotests/src/com/android/settings/biometrics/face/FaceEnrollTest.kt
+++ b/tests/robotests/src/com/android/settings/biometrics/face/FaceEnrollTest.kt
@@ -17,14 +17,17 @@
 package com.android.settings.biometrics.face
 
 import android.app.Activity
+import android.app.PendingIntent
 import android.content.Intent
+import androidx.test.core.app.ApplicationProvider
+import com.android.settings.biometrics.MultiBiometricEnrollHelper.EXTRA_ENROLL_AFTER_FACE
 import com.android.settings.overlay.FeatureFactory
 import com.android.settings.testutils.FakeFeatureFactory
 import com.google.common.truth.Truth.assertThat
 import org.junit.Before
 import org.junit.Test
 import org.junit.runner.RunWith
-import org.mockito.Mockito.`when` as whenever
+import org.mockito.kotlin.whenever
 import org.robolectric.Robolectric
 import org.robolectric.RobolectricTestRunner
 import org.robolectric.Shadows
@@ -36,10 +39,6 @@ class FaceEnrollTest {
     private companion object {
         const val INTENT_KEY = "testKey"
         const val INTENT_VALUE = "testValue"
-        val INTENT = Intent().apply {
-            putExtra(INTENT_KEY, INTENT_VALUE)
-            putExtra(Intent.EXTRA_USER_ID, 11)
-        }
     }
 
     private val activityProvider = FaceEnrollActivityClassProvider()
@@ -55,7 +54,21 @@ class FaceEnrollTest {
         activityClass: Class<out FaceEnroll>,
         launchedFrom: String? = null,
     ): FaceEnroll {
-        val activityController = Robolectric.buildActivity(activityClass, INTENT)
+        val activityIntent =
+            Intent().apply {
+                putExtra(INTENT_KEY, INTENT_VALUE)
+                putExtra(Intent.EXTRA_USER_ID, 11)
+                putExtra(
+                    EXTRA_ENROLL_AFTER_FACE,
+                    PendingIntent.getActivity(
+                        ApplicationProvider.getApplicationContext()!!,
+                        1,
+                        Intent(),
+                        0,
+                    ),
+                )
+            }
+        val activityController = Robolectric.buildActivity(activityClass, activityIntent)
         activityController.get().launchedFromProvider = {
             launchedFrom ?: activityController.get().packageName
         }
@@ -68,23 +81,22 @@ class FaceEnrollTest {
 
         val nextActivityIntent = activity.asShadow().nextStartedActivity
         assertThat(nextActivityIntent.component!!.className).isEqualTo(activityProvider.next.name)
-        assertThat(nextActivityIntent.extras!!.size()).isEqualTo(2)
+        assertThat(nextActivityIntent.extras!!.size()).isEqualTo(3)
         assertThat(nextActivityIntent.getStringExtra(INTENT_KEY)).isEqualTo(INTENT_VALUE)
         assertThat(nextActivityIntent.hasExtra(Intent.EXTRA_USER_ID)).isTrue()
+        assertThat(nextActivityIntent.hasExtra(EXTRA_ENROLL_AFTER_FACE)).isTrue()
     }
 
     @Test
-    fun testFinishAndLaunchDefaultActivity_fromExternal_dropUser() {
-        val activity = setupActivity(
-            FaceEnroll::class.java,
-            launchedFrom = "com.foo.bar"
-        )
+    fun testFinishAndLaunchDefaultActivity_fromExternal_dropUserAndDropEnrollAfterFace() {
+        val activity = setupActivity(FaceEnroll::class.java, launchedFrom = "com.foo.bar")
 
         val nextActivityIntent = activity.asShadow().nextStartedActivity
         assertThat(nextActivityIntent.component!!.className).isEqualTo(activityProvider.next.name)
         assertThat(nextActivityIntent.extras!!.size()).isEqualTo(1)
         assertThat(nextActivityIntent.getStringExtra(INTENT_KEY)).isEqualTo(INTENT_VALUE)
         assertThat(nextActivityIntent.hasExtra(Intent.EXTRA_USER_ID)).isFalse()
+        assertThat(nextActivityIntent.hasExtra(EXTRA_ENROLL_AFTER_FACE)).isFalse()
     }
 }
 
-- 
2.55.0.897.gb25b4bd76c-goog

